### PR TITLE
improve setPassword UX

### PR DESCRIPTION
**Pull request by anthony-hull**
### Details
This change replaces the set password form with the API error if the magic link has expired. 
This PR still needs a translation adding.
You need to wait a long time for the code to expire (or maybe someone with BE access to work around this?) so make sure you create a new account now! I'm opening the PR now to get eyes on it as there is a long lead time if you don't have a work around.
More testing screenshots to follow after PR is finalised.

### Fixed Issues

$ https://github.com/Expensify/App/issues/3189
$ https://github.com/Expensify/App/issues/6890

#### User signs up 
1. Create new account
2. Click link
3. Check API call for verify email ran
4. Set a new password with the form
5.  Check app loads in
--------
#### User visits page and returns later
1. Create new account
2. Click link
3. Refresh page
4. Set a new password with the form
5. Check app loads in
--------
#### User signs up and clicks link on a new browser
Testing steps for: https://github.com/Expensify/App/issues/6890
1. Create new account
2. Go to dev tools and clear local storage. (or open a new incognito session)
3. Click link
4. Set a new password with the form
5.  Check app loads in
--------
#### User visits page and returns later after authToken has expired
1. Create new account
2. Click link
3. Delete half the characters in the local storage key: userSignUp.authToken
4. set a new password with the form
5. Check the message informing the user of the failed operation (and a new link has been sent) replaces the form and the user has been navigated to the root of the app.
6. Check emails for a new link
7. Click link
8. Set a new password with the form
9. Check app loads in
--------
##### Ideal Flow
1. Create new account
2. Click link
3. Open  a new cognito session
4. Paste url into cognito session
5. Check the form is replaced by the error message to check email for a new link and the user has been navigated to the root of the app.
6. Check emails for a new link
7. Click link
8. Set a new password with the form
9. Check app loads in

##### Current Flow
Unfortunately I don't know if the magic key has been consumed at step 4. I can only know that when I try to use it with SetPassword. So these steps have been changed to:
1. Create new account
2. Click link
3. Open  a new cognito session
4. Paste url into cognito session
5.  Set a new password with the form
5. Check the form is replaced by the error message to check email for a new link and the user has been navigated to the root of the app.
6. Check emails for a new link
_Email currently is the URL for old site._
7. Click link
8. Set a new password with the form
9. Check app loads in
--------
#### User visits page on a slow connection
1. Create new account
2. Set network speed to slow 3g
3. Click link
4. See form button is locked and verifying message appears (be quick as it's waiting for the API.verifyEmail call to finish)
--------
#### User visits page via a link that has expired
1. Create a new account
2. Wait 7 days
3. Click link
4. Try to set a valid password
5. See the error replaces the form telling the user to check their emails for a new link. user is not navigated anywhere.
6. A new link should be waiting in your email inbox 

#### User visits page via a valid link
1. Reset Password on an account
2. Click link
3. Try to set a valid password
4. App should load in normal
5. Log out
6. Try to log in again with new password
7. Check app loads in properly

#### User visits page via a link that has expired
1. Reset Password on an account
2. Wait 2 days
3. Click link
4. Try to set a valid password
5. See the error replaces the form telling the user to check their emails for a new link and the user has been navigated to the root of the app.
6. A new link should be waiting in your email inbox.

#### User resets password, clicks link and returns later on another browser
1. Reset Password on an account
2. Click link
3. Open  a new cognito session
4. Paste url into cognito session
5. Try to set a valid password
6. App should load in normal

### Tested On

- [X] Web
- [ ] Mobile Web
- [ ] Desktop
- [ ] iOS
- [ ] Android

#### Web

https://user-images.githubusercontent.com/47184118/144764155-33aea024-e0c1-4e71-a9f4-bf1cadf0bb49.mp4

### CODE REVIEW

**Comment by AndrewGable** [comment]
Would easier test steps be:

1. Create a new account
3. Click validate link
4. Set a password
5. Log out
6. **Click the same validate link again**
4. Try to set a valid password
5. See the error replaces the form telling the user to check their emails for a new link
6. A new link should be waiting in your email inbox

But this brings up the question, is the only way to get this error if the link is expired? If it's already been used once, why would we send it again?

**Comment by anthony-hull** [comment]
This is a good point. If it has already been used, we should inform the user what went wrong and how to resolve it. This would need a distinction in the API response between these two failure states.

We should also move the user to the login screen somehow. 

Either putting the message on the login screen and move them there automatically or keeping them on this page with the message and giving them a button to redirect to the login screen.

**Comment by AndrewGable** [comment]
@roryabraham thoughts on the above since you worked on the API?

----

@anthony-hull here is the Spanish translation:

> This set password link is invalid or has expired. A new one is waiting for you in your email inbox!
El enlace para configurar tu contraseña ha expirado. Te hemos enviado un nuevo enlace a tu correo.

**Review by roryabraham — COMMENTED** [review]
This still requires a user to enter a new password before they find out that the `validateCode` is expired. I think this should be the goal:

- When a user lands on the `SetPasswordPage`, immediately use `API.validateEmail` to attempt to validate the account.
- If we get a `401 ValidateCode expired`, the validate code is expired and we show the error.
- If we get a `405 Account already validated`, the validate code is valid and we can show the `setPassword` form. Once filled out, call `API.setPassword` with the validate code to set the password.
- If the account validation succeeds with a `200` status code, show the set password form. Once filled out, call the `changePasswordAndSignIn` action.

I'll actually need to make one more small back-end change to make this flow work out, but what do you think @anthony-hull and @AndrewGable? This should prevent the user from ever having to enter a password, only to find that their validation link has expired.

**Comment by roryabraham** [comment]
Minor back-end PR to ensure that the `401 ValidateCode expired` error will be thrown before `405 Account already validated`, which will mean that receiving `405 Account already validated` implies that the validateCode is not expired. 

https://github.com/Expensify/Auth/pull/6205

**Comment by AndrewGable** [comment]
@anthony-hull - FYI [this change](https://github.com/Expensify/App/pull/6587#issuecomment-989218813) was deployed if you want to update the PR. Thanks!

**Comment by anthony-hull** [comment]
Sure! Thank you. I'll be back from the sun on Tuesday 😀 I'll work on this with my other issues from then!

**Comment by AndrewGable** [comment]
Enjoy the sun @anthony-hull 😎

**Comment by anthony-hull** [comment]
I've come across a case where if you click the link in a new incognito window, the user's email isn't present in local storage. I need this to request a new link in case of not having an authToken set or an expired token.

Could we change the API to also accept the accountID?

**Comment by roryabraham** [comment]
> I've come across a case where if you click the link in a new incognito window, the user's email isn't present in local storage. I need this to request a new link in case of not having an authToken set or an expired token.

Sorry, I'm having difficulty following the case you're seeking to address. Can you please write detailed reproduction steps for this edge case to help me understand?

**Comment by anthony-hull** [comment]
1. Create new account
2. Click link
3. Open link again in another browser or incognito


In this case I wouldn't have the users email availible as it's only set while signing up for the account. The cognito session doesn't have it set. I need it to  request a new link to exchange for an authToken

**Comment by roryabraham** [comment]
Yeah, I personally don't think we need to address that edge case, can we just default to the login page if the email is not set for some reason?

**Comment by AndrewGable** [comment]
I agree 👍

**Comment by anthony-hull** [comment]
sounds good to me

**Comment by anthony-hull** [comment]
It has been brought to my attention that there is another PR which is changing how malformed tokens are handled on this page:
https://github.com/Expensify/App/pull/5150
Although it doesn't cover the UX change we agreed of validating the code on page load. I think we would still get value from implementing this change after the other PR has merged. (it is close)

**Comment by chiragsalian** [comment]
Heads up, https://github.com/Expensify/App/pull/5150 has been merged.

**Comment by anthony-hull** [comment]
perfect thank you! I will start back on this today.

**Comment by anthony-hull** [comment]
@chiragsalian [the method for getting the email from the validateEmail](https://github.com/Expensify/App/issues/6890#issuecomment-1003154723) is actually a response from API.SetPassword which I wouldn't call in componentDidMount() as I don't know what the password would be yet. 

I have two options, redirect to nav root like said above or somehow get the email (I could call the SetPassword endpoint with a bad password and token but this is a bit hacky)

**Comment by anthony-hull** [comment]
When I have a user who lands on this page with a magic code that has already been used, I hit the ValidateEmail endpoint and get an already validated error. 
I then don't know that the magic key has already been used. The user could have landed on this page from the new user UX or the resetPassword UX. So I try to use it on the Session.setPassword endpoint. If the magic key hasn't been used before, this will work. But if it has been used before, I will get an error:

{
    "code": 666,
    "jsonCode": 404,
    "type": "Expensify\\Libs\\Error\\ExpError",
    "UUID": "4a16d616-589d-47a2-9e3a-50729c33520a",
    "message": "Validate code is likely expired",
    "title": "Validate code failed",
    "data": {
        "email": "hullo2u+21@gmail.com"
    },
    "htmlMessage": "",
    "requestID": "6caed3d49de40091-DEN"
}

A new reset email is sent out but the URL is for the old site. Also the code and Json code here doesn't match, is that a bug?
I would like to be able to tell the user a new email is waiting but the email isn't the right one.

**Comment by anthony-hull** [comment]
I would like to confirm if we can change the API to send the new dot link for this situation or not as my flow currently depends on it :)

**Comment by AndrewGable** [comment]
@roryabraham I think you might be best to answer this one?

**Comment by roryabraham** [comment]
> Also the code and Json code here doesn't match, is that a bug?

Nope, that's expected. You should use the `jsonCode` as the source of truth 😄 

> A new reset email is sent out but the URL is for the old site.

Ah yeah, that seems like a bug. I can investigate.

> A new reset email is sent out ... but the email isn't the right one.

Is this a separate bug? Or is the email sent to the correct address with the wrong site URL?

**Comment by anthony-hull** [comment]
> Is this a separate bug?

No the same as above :)

**Comment by anthony-hull** [comment]
Is a malformed magic token treated the same as an expired one?
Otherwise my expired user journeys would be a bit of a pain to test for the QAs

> Ah yeah, that seems like a bug. I can investigate.
Is there an issue open for this?

**Comment by anthony-hull** [comment]
how do you want proof of testing for this? I've just finished running through all the tests again and they passed. 

Do you want videos of each one being done? it would be very long and boring videos :P or a lot of editing :( each test takes 1-2 min.

Do I need to do all the tests on all platforms? Or what strategy do you suggest?

I also have a few translations that need Spanish

File: src/pages/SetPasswordPage.js [hunk 1]
```diff
@@ -97,29 +118,29 @@ class SetPasswordPage extends Component {
...
+                                    style={[styles.mb2]}
+                                    text={buttonText}
+                                    isLoading={this.props.account.loading || this.props.userSignUp.isValidatingEmail}
+                                    onPress={this.validateAndSubmitForm}
+                                    isDisabled={!this.state.isFormValid}
+                                />
+                            </View>
+                        </>
                     )}
+                    {!_.isEmpty(error) && <Text>{error}</Text>}
```
Comment (roryabraham) [inline src/pages/SetPasswordPage.js:143]: NAB/Up to you, but since we're rendering one block or the other I might suggest using a ternary instead of `&&`

**Review by roryabraham — CHANGES_REQUESTED** [review]
I asked for a translation in slack, I'll post it here when I have it.

> Is a malformed magic token treated the same as an expired one?

I believe so, yes.

> A new reset email is sent out but the URL is for the old site.

Specifically what were the reproduction steps for this? Knowing exactly which testing flow causes this will help with my investigation.

> Is there an issue open for this?

No open issue, but if you create consistent reproduction steps I should be able to get it sorted.

File: src/pages/SetPasswordPage.js [hunk 2]
```diff
@@ -75,20 +89,27 @@ class SetPasswordPage extends Component {
...
-     */
-    validateAndSubmitForm() {
+    componentDidMount() {
         const accountID = lodashGet(this.props.route.params, 'accountID', '');
         const validateCode = lodashGet(this.props.route.params, 'validateCode', '');
-        if (!this.state.isFormValid) {
+        if (this.props.userSignUp.authToken) {
             return;
         }
+        Session.validateEmail(accountID, validateCode, this.props.credentials.login, this.props.userSignUp.authToken);
```
[the remark this answers is not in the document]
Comment (anthony-hull) [reply src/pages/SetPasswordPage.js:98]: Ahh thanks, I was doing more in this before I cleaned it up. Missed those!

**Comment by anthony-hull** [comment]
> No open issue, but if you create consistent reproduction steps I should be able to get it sorted.

If you follow the _current flow_ steps of _User visits page and returns later on another browser_ you will get the email at step 7. 

If you compare the ideal flow and current flow, I think there is the potential of improving the UX slightly if the API response can tell us that the validate code has been used up as well as telling us the account has already been validated. we could save the user from entering a password only to be sent to the root nav and told to start again.
What do you think?

**Comment by anthony-hull** [comment]
> > I also have a few translations that need Spanish
> 
> `Verifying account` -> `Verificando cuenta`

Thanks, sorry I should have extracted out the keys I was asking about, also just need:
`passwordNotSet: 'We were unable to set your new password correctly. We have sent you a new password link to try again.',`
updated:
`passwordNotSet: 'No pudimos establecer to contaseña correctamente.',`
We have an extra sentence on the end now.

**Comment by roryabraham** [comment]
Okay, I'll get those translations

**Comment by roryabraham** [comment]
Okay, the english copy should actually be:

`We were unable to set your new password. We have sent you a new password link to try again.` (i.e: remove "correctly")

y en español: `No pudimos cambiar tu clave. Te hemos enviado un nuevo enlace para que intentes cambiar la clave nuevamente`

**Comment by anthony-hull** [comment]
Updated

**Review by AndrewGable — APPROVED** [review]

**Comment by AndrewGable** [comment]
All yours @roryabraham

**Review by roryabraham — APPROVED** [review]

**Comment by anthony-hull** [comment]
> No open issue, but if you create consistent reproduction steps I should be able to get it sorted.

hey @roryabraham were you able to follow my reproduction steps for this?

1. Create new account
2. Click link
3. Open a new cognito session
4. Paste url into cognito session
5. Set a new password with the form
6. Check the form is replaced by the error message to check email for a new link and the user has been navigated to the root of the app.
7. Check emails for a new link
 _Email currently is the URL for old site._
9. Click link
10. Set a new password with the form
11. Check app loads in

**Pull request merged**